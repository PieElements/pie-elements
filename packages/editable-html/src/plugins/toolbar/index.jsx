import React from 'react';
import Toolbar from './toolbar';
import debug from 'debug';
const log = debug('editable-html:toolbar');

// export class Toolbar extends React.Component {

//   render() {
//     const { value, plugins } = this.props;
//     log('editor:', value);

//     return <div>
//       <div>TOOLBAR {value.isFocused ? 'focus' : ''}</div>
//     </div>
//   }
// }


export default function ToolbarPlugin(opts) {
  return {
    renderEditor: (props) => (
      <div>
        <div>{props.children}</div>
        {props.value.isFocused && <Toolbar value={props.value} />}
      </div>
    )
  }
}