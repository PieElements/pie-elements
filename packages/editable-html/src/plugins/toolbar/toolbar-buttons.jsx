import React from 'react';
import debug from 'debug';
import injectSheet from 'react-jss';
import classNames from 'classnames';

const styles = () => ({
  button: {
    color: 'grey',
    display: 'inline-flex',
    padding: '2px',
    '& :hover': {
      color: 'black'
    }
  },
  active: {
    color: 'black'
  }
});

const log = debug('pie-elements:editable-html:raw-button');

export class RawButton extends React.Component {

  constructor(props) {
    super(props)

    this.onClick = (e) => {
      log('[onClick]');
      e.preventDefault();
      const { onClick } = this.props;
      onClick(e)
    }
  }

  render() {
    const { classes, children, active } = this.props;
    const names = classNames(classes.button, active && classes.active);

    return (
      <div
        className={names}
        onMouseDown={this.onClick}>
        {children}
      </div>
    );
  }
}

export const Button = injectSheet(styles())(RawButton);


export class RawMarkButton extends React.Component {
  constructor(props) {
    super(props);

    this.onToggle = (e) => {
      e.preventDefault();
      this.props.onToggle(this.props.mark);
    };
  }

  render() {
    const { classes, children, active } = this.props;
    const names = classNames(classes.button, active && classes.active);
    return (
      <span
        className={names}
        onMouseDown={this.onToggle}>
        {children}
      </span>
    );
  }
}

export const MarkButton = injectSheet(styles())(RawMarkButton);