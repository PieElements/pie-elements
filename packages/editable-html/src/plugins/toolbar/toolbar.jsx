import { Button, MarkButton } from './toolbar-buttons';

import Bold from 'material-ui-icons/FormatBold';
import Check from 'material-ui-icons/Check';
import Code from 'material-ui-icons/Code';
import Functions from 'material-ui-icons/Functions';
import Image from 'material-ui-icons/Image';
import Italic from 'material-ui-icons/FormatItalic';
import PropTypes from 'prop-types';
import React from 'react';
import Strikethrough from 'material-ui-icons/FormatStrikethrough';
import Underlined from 'material-ui-icons/FormatUnderlined';
import debug from 'debug';
import injectSheet from 'react-jss';

const log = debug('editable-html:plugins:toolbar');

const toolbarStyle = {
  toolbar: {
    position: 'absolute',
    zIndex: 10,
    display: 'flex',
    cursor: 'pointer',
    justifyContent: 'space-between',
    background: 'var(--editable-html-toolbar-bg, #efefef)',
    margin: '0px',
    padding: '2px',
    width: '100%',
    boxShadow: '0px 1px 5px 0px rgba(0, 0, 0, 0.2), 0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 3px 1px -2px rgba(0, 0, 0, 0.12)',
    boxSizing: 'border-box'
  }
}

const ToolbarButton = (props) => {

  const hasMark = (type) => {
    const { value } = props;
    return value.marks.some(mark => mark.type == type)
  }

  if (props.isMark) {
    const isActive = hasMark(props.type);
    return <MarkButton
      active={isActive}
      label={props.type}
      onToggle={() => {
        const c = props.onToggle(props.value.change())
        props.onChange(c);
      }}
      mark={props.type}
    >{props.icon}</MarkButton>
  } else {
    return <Button
      onClick={props.onClick}>{props.icon}</Button>
  }
}

class RawToolbar extends React.Component {

  constructor(props) {
    super(props);
    this.hasMark = (type) => {
      const { value } = this.props;
      return value.marks.some(mark => mark.type == type)
    }

    this.hasBlock = (type) => {
      const { value } = this.props;
      return value.blocks.some(node => node.type == type)
    }
  }

  onToggle = (plugin) => {
    const { value, onChange } = this.props;

    if (!plugin.onToggle) return;

    const change = plugin.onToggle(value.change())
    onChange(change);
  }

  render() {
    const {
      classes,
      onToggleMark,
      onImageClick,
      onInsertMath,
      onDone,
      zIndex,
      onFocus,
      onBlur,
      plugins,
      value,
      onChange } = this.props;

    const toolbarPlugins = plugins.filter(p => p.toolbar).map(p => p.toolbar);

    const style = zIndex ? { zIndex } : {};

    return (
      <div className={classes.toolbar}
        style={style}
        onFocus={onFocus}
        onBlur={onBlur}>
        <div className={classes.inline}>
          {toolbarPlugins.map((p, index) => {
            return <ToolbarButton
              {...p}
              key={index}
              value={value}
              onChange={onChange} />
            // const isActive = this.hasMark(p.type);
            // return <MarkButton
            //   key={p.type}
            //   active={isActive}
            //   label={p.type}
            //   onToggle={() => this.onToggle(p)}
            //   mark={p.type}
            // >{p.icon}</MarkButton>
          }
          )}
          {/* {onImageClick && <Button onClick={onImageClick}> <Image /></Button>}
          <Button onClick={onInsertMath}> <Functions /></Button> */}
        </div>
        <Button onClick={onDone}><Check /></Button>
      </div>
    );
  }
}

RawToolbar.propTypes = {
  zIndex: PropTypes.number,
  value: PropTypes.object.isRequired,
  plugins: PropTypes.array,
  onToggleMark: PropTypes.func.isRequired,
  onImageClick: PropTypes.func,
  onInsertMath: PropTypes.func.isRequired,
  onDone: PropTypes.func.isRequired
}

export default injectSheet(toolbarStyle)(RawToolbar);