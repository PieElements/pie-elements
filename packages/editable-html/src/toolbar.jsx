import { Button, MarkButton } from './toolbar-buttons';

import Bold from 'material-ui-icons/FormatBold';
import Check from 'material-ui-icons/Check';
import Functions from 'material-ui-icons/Functions';
import Image from 'material-ui-icons/Image';
import Italic from 'material-ui-icons/FormatItalic';
import React from 'react';
import Underlined from 'material-ui-icons/FormatUnderlined';
import injectSheet from 'react-jss';

const toolbarStyle = {
  root: {
    display: 'flex',
    cursor: 'pointer',
    justifyContent: 'space-between',
    background: 'var(--editable-html-toolbar-bg, #eeeeee)',
    borderTop: 'solid 1px #cccccc',
    margin: '0px',
    padding: '2px'
  }
}

var INLINE_STYLES = [
  { mark: 'b', label: 'Bold', icon: <Bold /> },
  { mark: 'i', label: 'Italic', icon: <Italic /> },
  { mark: 'u', label: 'Underline', icon: <Underlined /> }
];

class RawToolbar extends React.Component {

  constructor(props) {
    super(props);
    this.hasMark = (type) => {
      const { editorState } = this.props;
      return editorState.marks.some(mark => mark.type == type)
    }

    this.hasBlock = (type) => {
      const { editorState } = this.props;
      return editorState.blocks.some(node => node.type == type)
    }
  }

  render() {
    const {
      editorState,
      classes,
      onToggleMark,
      onImageClick,
      onInsertMath,
      onDone } = this.props;

    return (
      <div className={classes.root}>
        <div className={classes.inline}>
          {INLINE_STYLES.map(type => {
            const isActive = this.hasMark(type.mark);
            return <MarkButton
              key={type.label}
              active={isActive}
              label={type.label}
              onToggle={onToggleMark}
              mark={type.mark}
            >{type.icon}</MarkButton>
          }
          )}
          <Button onClick={onImageClick}> <Image /></Button>
          <Button onClick={onInsertMath}> <Functions /></Button>
        </div>
        <Button onClick={onDone}><Check /></Button>
      </div>
    );
  }
}

export default injectSheet(toolbarStyle)(RawToolbar);