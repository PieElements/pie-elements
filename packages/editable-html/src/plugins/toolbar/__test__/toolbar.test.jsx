import { Data } from 'slate';
import { RawToolbar } from '../toolbar';
import React from 'react';
import renderer from 'react-test-renderer';
import { stub } from 'sinon';

jest.mock('material-ui/IconButton', () => {
  return (props) => <div className={props.className} style={props.style} ariaLabel={props['aria-label']}></div>
})

const simpleObject = function () {
  const out = {}
  for (var a in arguments) {
    const value = arguments[a];
    out[value] = value;
  }
  return out;
}

describe('toolbar', () => {

  it('renders custom toolbar', () => {

    const node = {}

    const value = {
      isCollapsed: true,
      startKey: 1,
      document: {
        getClosestInline: stub().returns(node)
      }
    }

    const plugins = [
      {

      }
    ];
  });

  it('renders default toolbar', () => {
    const onDelete = stub();

    const classes = simpleObject('inline', 'toolbar', 'focused', 'shared', 'inline')

    const plugins = [
      {
        toolbar: {}
      }
    ];

    const tree = renderer
      .create(<RawToolbar
        plugins={plugins}
        classes={classes}
        value={{}}
        onDone={() => ({})} />, {
        createNodeMock: el => {
          if (el.type === 'img') {
            return {
              naturalWidth: 100,
              naturalHeight: 100
            }
          }
        }
      })
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
