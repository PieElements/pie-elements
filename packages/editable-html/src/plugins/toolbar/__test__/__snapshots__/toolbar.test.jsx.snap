// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toolbar renders correctly 1`] = `
<div
  className="toolbar"
  onClick={[Function]}
>
  <div
    className={undefined}
  >
    <div
      className="root-0-4"
      onMouseDown={[Function]}
    />
  </div>
  <div
    className="shared"
  >
    <div
      ariaLabel="Done"
      className={undefined}
      style={
        Object {
          "height": "28px",
          "width": "28px",
        }
      }
    />
  </div>
</div>
`;
